### PR TITLE
Upgrade wagmi to v1

### LINKED ISSUE #425: Failed to execute 'setItem' on 'Storage' (local storage exceeded)
The web3modal uses local storage for various things which sometimes runs out of space

### PR DESCRIPTION
**Pull request by mcoetzee**
Fixes #425 

According to [this issue](https://github.com/WalletConnect/web3modal/issues/1345), the problem should be resolved by updating to the latest `wagmi`.

### What does this change?
- upgrades `wagmi` to v1, which comes with `viem`
- upgrades Typescript to v5 (`viem` depends on Typescript v5)
- shows a warning message when `localStorage.setItem()` fails

### How did you action this task?
- adapters provided by `wagmi` (https://wagmi.sh/react/ethers-adapters) were used to convert the `viem` public and wallet clients to the `ethers` provider and signer.
- I don't have full confidence that the original issue is resolved, so I thought it would be a good idea to let the user know if/when it happens again and to give them the ability to quickly sort it out.

### Screenshots
Screenshot 2023-10-03 at 15 43 59

### CODE REVIEW

File: setup-jest.js [hunk 1]
```diff
@@ -0,0 +1,4 @@
+import { TextEncoder, TextDecoder } from 'util';
+
+global.TextEncoder = TextEncoder;
```
Comment (Siegrift) [inline setup-jest.js:3]: Mhh, can you explain why is this needed?

File: src/__test_utils__/mock-wagmi-client.ts [hunk 2]
```diff
@@ -1,51 +1,35 @@
...
-    privateKey: '0xac0974bec39a17e36ba4a6b4d238ff944bacb478cbed5efcae784d7bf4f2ff80',
-    balance: '10000000000000000000000',
-  },
-  {
-    // Account #1: 0x70997970C51812dc3A010C7d01b50e0d17dc79C8
-    privateKey: '0x59c6995e998f97a5a0044966f0945389dc9e86dae88c7a8412f4603b6b78690d',
-    balance: '10000000000000000000000',
-  },
-];
+export const getPublicClient = () => {
```
Comment (Siegrift) [inline src/__test_utils__/mock-wagmi-client.ts:29]: Do we have this viem migration tested somewhere? I am a bit paranoid because I am not that familiar with the library.

File: src/chain-data/adapters.ts
Comment (Siegrift) [inline src/chain-data/adapters.ts:1]: I really love that we can migrate gradually.

**Review by Siegrift — APPROVED** [review]
👍 LGTM

File: setup-jest.js (hunk 1 above)
Comment (mcoetzee) [reply setup-jest.js:3]: That's to get rid of these errors when running the tests:
```
    ReferenceError: TextEncoder is not defined

      1 | import { createContext, useState, useMemo, useContext, ReactNode } from 'react';
      2 | import { initialSettableChainData, initialChainData, SettableChainData } from './state';
    > 3 | import { useAccount, useNetwork } from 'wagmi';
        | ^
      4 | import { getDaoAddresses, updateNetworkName } from '../contracts';
      5 | import { useEthersProvider, useEthersSigner } from './adapters';
      6 | import { ethers } from 'ethers';

      at Object.<anonymous> (node_modules/viem/utils/encoding/toHex.ts:226:31)
      at Object.<anonymous> (node_modules/viem/utils/encoding/toBytes.ts:8:1)
      at Object.<anonymous> (node_modules/viem/utils/encoding/fromHex.ts:12:1)
      at Object.<anonymous> (node_modules/viem/chains/optimism/formatters.ts:4:1)
      at Object.<anonymous> (node_modules/viem/chains/definitions/base.ts:2:1)
      at Object.<anonymous> (node_modules/viem/chains/index.ts:10:1)
      at Object.<anonymous> (node_modules/@wagmi/core/dist/chunk-4E64EZ7V.js:2:1)
      at Object.<anonymous> (node_modules/@wagmi/core/dist/index.js:1:1)
      at Object.<anonymous> (node_modules/wagmi/dist/index.js:7:1)
      at Object.<anonymous> (src/chain-data/context.tsx:3:1)
      at Object.<anonymous> (src/chain-data/index.ts:2:1)
```

File: src/__test_utils__/mock-wagmi-client.ts (hunk 2 above)
Comment (mcoetzee) [reply src/__test_utils__/mock-wagmi-client.ts:29]: Yeah the Market has been running with these `viem` to `ethers` adapters since the end of June. And fyi incase it wasn't clear, these are the test utils that you are looking at. It is used here:
```ts
const wagmiClient = createMockClient();

export const render = (ui: ReactElement, options?: RenderOptions): RenderResult => {
  return baseRender(
    <BrowserRouter>
      <WagmiConfig config={wagmiClient}>{ui}</WagmiConfig>
    </BrowserRouter>,
    options
  );
};
```

Comment (Siegrift) [reply src/__test_utils__/mock-wagmi-client.ts:29]: I meant it in general, but the question came out when I was looking at those tests utils, so I put it there. Sorry for confusing you.
